app/src/main/java/org/thoughtcrime/securesms/providers/PartProvider.java

```diff
@@ -0,0 +1,170 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.thoughtcrime.securesms.providers;
+
+import android.content.ContentUris;
+import android.content.ContentValues;
+import android.content.UriMatcher;
+import android.database.Cursor;
+import android.net.Uri;
+import android.os.MemoryFile;
+import android.os.ParcelFileDescriptor;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+
+import org.signal.core.util.StreamUtil;
+import org.signal.core.util.logging.Log;
+import org.thoughtcrime.securesms.BuildConfig;
+import org.thoughtcrime.securesms.attachments.AttachmentId;
+import org.thoughtcrime.securesms.attachments.DatabaseAttachment;
+import org.thoughtcrime.securesms.database.DatabaseFactory;
+import org.thoughtcrime.securesms.mms.PartUriParser;
+import org.thoughtcrime.securesms.service.KeyCachingService;
+import org.thoughtcrime.securesms.util.MemoryFileUtil;
+import org.thoughtcrime.securesms.util.Util;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+
+public final class PartProvider extends BaseContentProvider {
+
+  private static final String TAG = Log.tag(PartProvider.class);
+
+  private static final String CONTENT_AUTHORITY  = BuildConfig.APPLICATION_ID + ".part";
+  private static final String CONTENT_URI_STRING = "content://" + CONTENT_AUTHORITY + "/part";
+  private static final Uri    CONTENT_URI        = Uri.parse(CONTENT_URI_STRING);
+  private static final int    SINGLE_ROW         = 1;
+
+  private static final UriMatcher uriMatcher;
+
+  static {
+    uriMatcher = new UriMatcher(UriMatcher.NO_MATCH);
+    uriMatcher.addURI(CONTENT_AUTHORITY, "part/*/#", SINGLE_ROW);
+  }
+
+  @Override
+  public boolean onCreate() {
+    Log.i(TAG, "onCreate()");
+    return true;
+  }
+
+  public static Uri getContentUri(AttachmentId attachmentId) {
+    Uri uri = Uri.withAppendedPath(CONTENT_URI, String.valueOf(attachmentId.getUniqueId()));
+    return ContentUris.withAppendedId(uri, attachmentId.getRowId());
+  }
+
+  @Override
+  public ParcelFileDescriptor openFile(@NonNull Uri uri, @NonNull String mode) throws FileNotFoundException {
+    Log.i(TAG, "openFile() called!");
+
+    if (KeyCachingService.isLocked(getContext())) {
+      Log.w(TAG, "masterSecret was null, abandoning.");
+      return null;
+    }
+
+    if (uriMatcher.match(uri) == SINGLE_ROW) {
+      Log.i(TAG, "Parting out a single row...");
+      try {
+        final PartUriParser partUri = new PartUriParser(uri);
+        return getParcelStreamForAttachment(partUri.getPartId());
+      } catch (IOException ioe) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)